### ISSUE TITLE
Cannot allocate memory in static TLS block

### ISSUE THREAD

**Issue by eikenb** [flox/flox#1341]
**Describe the bug:**
Running `dig` or `host` inside a flox environment I get the error..
```sh
$ dig github.com
dig: error while loading shared libraries: /lib/x86_64-linux-gnu/libjemalloc.so.2: cannot allocate memory in static TLS block
```
This just started happening and I've yet to find a cause in environments that were working fine. Outside the flox environments dig/host still work fine. It happens with brand new, empty environments. 

Digging around a bit I found [issue 2472](https://github.com/jemalloc/jemalloc/issues/2472) in jemalloc's repo. I found that flox sets LD_AUDIT is set and if I unset that environment variable I no longer get the error (dig works normally).

**Steps to reproduce:**
1. flox init
2. flox activate
3. dig github.com

Though I doubt you will be able to repro it as it started happening after a period of time and a lot of activity.

Result: error shown above

Expected: dig's normal output (dns lookup info)

**Screenshots/terminal output (if relevant):**

**flox Version (run `flox --version` if possible):**

1.0.3

**`uname -a` output:**

Linux hostname 6.1.0-18-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.76-1 (2024-02-01) x86_64 GNU/Linux

**Comment by tomberek** [comment]
Can you post a copy of your manifest? What other things are installed?

**Comment by eikenb** [comment]
Thanks for the fast response.

These are in empty/new environments. So nothing is installed and they have the default manifest.

Globally I only have a few environments setup. There is only one with anything installed at this point. It gets the error like the others. Here is its manifest...

$ flox list --config
```yaml
#
# This is a flox environment manifest.
# Visit flox.dev/docs/concepts/manifest/
# or see flox-edit(1) for more information
#

# List packages you wish to install in your environment inside
# the 'install' section

[install]
neovim.pkg-path = "neovim"
helix.pkg-path = "helix"
podman.pkg-path = "podman"
# hello.pkg-path = "hello"
# nodejs = { version = "^18.4.2", pkg-path = "nodejs_18" }

# Set an environment variable.
# These variables may not reference once another
# and are added to the environment without expansion.

[vars]
# message = "Howdy"
# pass-in = "$some-env-var"

# An activation hook will be run when entering the environment.
# You can define one in the 'hook' table inline via the 'script' field,
# or provide a file using the 'file' field.
# If 'file' is provided, the path should be relative to this file.
# If both 'file' and 'script' are provided,
# only the hook defined in 'file' will be run .

[hook]
# script = """
#   echo "it's gettin flox in here";
# """

# An environment that works on one system is guaranteed to work on the
# same type of system,
# but other systems may not have the same packages available, etc.
# In order to use the environment on a system you must explicitly
# add it to this list.
[options]
systems = ["x86_64-linux"]
```

**Comment by eikenb** [comment]
One possibly interesting thing I just discovered is that if I load flox into the current shell via `eval "$(flox activate)"` I **do not** get the error.

**Comment by eikenb** [comment]
In relation to the above, I checked and when I load it into the current shell via `eval` it **does not** have LD_AUDIT set.

**Comment by tomberek** [comment]
So far we've seen this with `redis-cli` which uses a non-standard jemalloc. Is it possible you have an LD_PRELOAD used to replace the normal allocator with jemalloc, and then thereafter we have a conflict? I believe this is how jemalloc is often used as the system-level to replace the default allocator at runtime.

- Do you have LD_PRELOAD set to something like this? `LD_PRELOAD=${JEMALLOC_PATH}/lib/libjemalloc.so.1`
- Can you confirm a jemalloc installation at the system level?
- Do you still get an error if you have LD_AUDIT set, but unset LD_PRELOAD?

**Comment by eikenb** [comment]
> * Do you have LD_PRELOAD set to something like this? `LD_PRELOAD=${JEMALLOC_PATH}/lib/libjemalloc.so.1`

No LD_PRELOADs in my current setup. It seems to be linked against jemalloc by Debian. Dig here is part of the standard bind9-dnsutils package and is linked against libjemalloc...

```sh
$ ldd /usr/bin/dig | grep jemalloc
	libjemalloc.so.2 => /lib/x86_64-linux-gnu/libjemalloc.so.2 (0x00007fdf3ba00000)
```

> * Can you confirm a jemalloc installation at the system level?

It is...
```sh
$ dpkg -l | grep libjemalloc2
ii  libjemalloc2:amd64                               5.3.0-1                                   amd64        general-purpose scalable concurrent malloc(3) implementation
```

> 
> * Do you still get an error if you have LD_AUDIT set, but unset LD_PRELOAD?

LD_PRELOAD is not set in the flox environment with this issue, nor do I have it set anywhere.

```sh
$ env | grep LD_
LD_FLOXLIB=/nix/store/vxdp9df5hw0lsd199blxvwpns9pf8xsn-flox-pkgdb-0.1.0/lib/ld-floxlib.so
LD_AUDIT=/nix/store/vxdp9df5hw0lsd199blxvwpns9pf8xsn-flox-pkgdb-0.1.0/lib/ld-floxlib.so
```

**Comment by eikenb** [comment]
> Though I doubt you will be able to repro it as it started happening after a period of time and a lot of activity.

FYI this isn't the case. It seems like it is the combo of LT_AUDIT + jemalloc using commands. And if I force other commands (eg. curl) to use jemalloc w/ LD_PRELOAD it gives the error.

**Comment by tomberek** [comment]
Yep, I was able to reproduce. I'll see how minimal I can make an example and see if there is a fix. In the meantime, one solution is to add "unset LD_AUDIT" to your hooks; until we have fix ready. Thanks for the report.

**Comment by eikenb** [comment]
I used the eval technique to load it into the current shell instead as it doesn't set LD_AUDIT.

**Comment by ghudgins** [comment]
need to create a bug upstream 

solution ideas:
 - create wrappers that unset LD audit just for these apps? this would need to modify nixpkgs in the catalog. we could include this in catalog. we could include these modifications in the buildEnv fetcher in the client which could bypass issues with revs on the catalog server
 - bug upstream

**Comment by mkenigs** [comment]
Discussion summary:
- The workaround for now is is `unset LD_AUDIT` for anything that depends on jemalloc
- We'll investigate a fix upstream

**Comment by mkenigs** [comment]
Status update on this one:

I opened https://sourceware.org/bugzilla/show_bug.cgi?id=31991 against glibc. It sounds like glibc should possibly reserve more TLS than it currently does

Even if that is fixed, there may be a separate bug causing jemalloc to segfault when using LD_AUDIT

**Comment by mkenigs** [comment]
This also impacts `fd` per @stealthybox

**Comment by mkenigs** [comment]
And `biome` per @rawkode

**Comment by mkenigs** [comment]
We use it to allow finding libraries in `.flox/run/*/lib`. This is useful for development if you need to compile against libraries installed into your environment. If you don't have libs installed into your environment, it's totally okay to unset it.

We may at some point add a setting for environments to be run in development mode vs runtime mode. Runtime mode would essentially only be for running binaries in PATH. Development mode would be for situations where you need libraries working. That could be a C library, which we expose via LD_AUDIT, or a Python module, which we expose via setting PYTHONPATH. That touches some on https://github.com/flox/flox/issues/1268. Does that make sense?

**Comment by eikenb** [comment]
Makes sense to me. Thank you for the clarification!

**Comment by sciyoshi** [comment]
I've found a better workaround to this thanks to [this comment](https://sourceware.org/bugzilla/show_bug.cgi?id=31991#c1) on an issue that was opened upstream against glibc. Setting a tunable to increase the static TLS size is the fix:

`export GLIBC_TUNABLES=glibc.rtld.optional_static_tls=10000`

This is better because it avoids needing to unset `LD_AUDIT` where there's some situations where that might still be necessary.

**Comment by tomberek** [comment]
Fixed-by: [flox/flox#2976](https://github.com/flox/flox/issues/2976)
